recommendation/src/main/java/com/yas/recommendation/topology/ProductTopology.java

```diff
@@ -0,0 +1,359 @@
+                        kafkaTopicConfig.category(),
+                        Consumed.with(getSerde(KeyDto.class), getMessageDtoSerde(CategoryDto.class)))
+                .selectKey((key, value) -> key.getId())
+                .mapValues(this::extractModelFromMessage)
+                .toTable(
+                        createMaterializedStore(
+                                CATEGORY_MATERIALIZED_VIEW, Serdes.Long(), getSerde(CategoryDto.class)));
+    }
+
+    private KTable<Long, ProductCategoryDto> createProductCategoryTable(StreamsBuilder streamsBuilder) {
+        return streamsBuilder.stream(
+                        kafkaTopicConfig.productCategory(),
+                        Consumed.with(getSerde(KeyDto.class), getMessageDtoSerde(ProductCategoryDto.class)))
+                .selectKey((key, value) -> key.getId())
+                .mapValues(this::extractModelFromMessage)
+                .toTable(
+                        createMaterializedStore(
+                                PRODUCT_CATEGORY_MATERIALIZED_VIEW, Serdes.Long(), getSerde(ProductCategoryDto.class)));
+    }
+
+    private KTable<Long, ProductCategoryDto> filterNonNullProductCategory(
+            KTable<Long, ProductCategoryDto> productCategoryTable) {
+        return productCategoryTable.toStream()
+                .filter((key, value) -> value != null)
+                .toTable(
+                        createMaterializedStore(
+                                PRODUCT_CATEGORY_NON_NULL_MATERIALIZED_VIEW,
+                                Serdes.Long(),
+                                getSerde(ProductCategoryDto.class)));
+    }
+
+    private KTable<Long, ProductCategoryDto> enrichProductCategoryWithDetails(
+            KTable<Long, ProductCategoryDto> productCategoryTable, KTable<Long, CategoryDto> categoryTable) {
+        return productCategoryTable.leftJoin(
+                categoryTable,
+                ProductCategoryDto::getCategoryId,
+                (productCategory, category) -> {
+                    productCategory.setCategoryName(category.getName());
+                    return productCategory;
+                },
+                createMaterializedStore(
+                        PRODUCT_CATEGORY_WITH_NAME_MATERIALIZED_VIEW,
+                        Serdes.Long(),
+                        getSerde(ProductCategoryDto.class)));
+    }
+
+    private KTable<Long, AggregationDto<CategoryDto>> aggregateProductCategoryDetails(
+            KTable<Long, ProductCategoryDto> productCategoryDetailTable,
+            KTable<Long, ProductCategoryDto> nonNullProductCategoryTable) {
+        return productCategoryDetailTable.toStream()
+                .mapValues(v -> v != null ? v : new ProductCategoryDto(-1L))
+                .join(
+                        nonNullProductCategoryTable,
+                        (productCategoryDetail, nonNullProductCategoryDetail) -> {
+                            if (productCategoryDetail.getId().equals(-1L)) {
```

Comment:
Using this approach to check for deleted cases doesn’t make sense. In the future, if there is master data with `id: -1`, will it be identified as deleted?

We should define a flag in `BaseMetaDataDto` to serve all cases.